pkg/virtual/generic/apiserver/apiserver.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2021 The KCP Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"net/http"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/registry/rest"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+
+	"github.com/kcp-dev/kcp/pkg/virtual/generic/builders"
+)
+
+type virtualNamespaceNameKeyType string
+
+const VirtualNamespaceNameKey virtualNamespaceNameKeyType = "VirtualWorkspaceName"
+
+type ExtraConfig struct {
+	builders.SharedExtraConfig
+	AdditionalConfig interface{}
```

Comment:
not sure we want this kind of nesting. It breaks the pattern of how we extend configs.

Reply:
We need some sort of composition so that the ExtraConfig of a given APIGroup API server implementation (for `workspaces` for example) can have additional fields. Go provides neither generics, nor inheritance ;-)
I'm very open to proposals in this regard.

Reply:
Changed